File: medium/0347_top_k_frequent_elements/solution.py
import heapq
from typing import List


class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        frequencies_map = {}

        for num in nums:
            frequencies_map[num] = frequencies_map.get(num, 0) + 1

        heap = [(-frequency, num) for num, frequency in frequencies_map.items()]
        heapq.heapify(heap)

        return [heapq.heappop(heap)[1] for _ in range(k)]

File: medium/0347_top_k_frequent_elements/solution.cpp
#include <queue>
#include <unordered_map>
#include <vector>

class Solution {
public:
  std::vector<int> topKFrequent(std::vector<int> &nums, int k) {
    std::unordered_map<int, int> numsCounter;
    std::priority_queue<std::pair<int, int>> pq;
    std::vector<int> topKFreqientNums;

    for (const int &num : nums) {
      if (numsCounter.find(num) == numsCounter.end())
        numsCounter[num] = 1;
      else
        numsCounter[num]++;
    }

    for (std::unordered_map<int, int>::iterator it = numsCounter.begin();
         it != numsCounter.end(); ++it)
      pq.push(std::pair<int, int>(it->second, it->first));

    for (int _ = 0; _ < k; ++_) {
      topKFreqientNums.push_back(pq.top().second);
      pq.pop();
    }

    return topKFreqientNums;
  }
};
